nntrainer/dataset/databuffer.h

```diff
@@ -201,101 +184,48 @@ class DataBuffer : public ml::train::Dataset {
-  std::vector<std::vector<float>> train_data;
-  std::vector<std::vector<float>> train_data_label;
-  std::vector<std::vector<float>> val_data;
-  std::vector<std::vector<float>> val_data_label;
-  std::vector<std::vector<float>> test_data;
-  std::vector<std::vector<float>> test_data_label;
-
-  /**
-   * @brief     feature size
-   */
-  TensorDim input_dim;
-
-  /**
-   * @brief     number of class
-   */
-  unsigned int class_num;
-
-  /**
-   * @brief     number of remain data for each data queue
-   */
-  unsigned int cur_train_bufsize;
-  unsigned int cur_val_bufsize;
-  unsigned int cur_test_bufsize;
-
-  /**
-   * @brief     queue size for each data set
-   */
-  unsigned int train_bufsize;
-  unsigned int val_bufsize;
-  unsigned int test_bufsize;
-
-  unsigned int max_train;
-  unsigned int max_val;
-  unsigned int max_test;
+  };
 
   /**
-   * @brief     remain data set size
+   * @brief Set the State And Notify to the condition variable waiting for it
+   *
+   * @param status status to change
    */
-  unsigned int rest_train;
-  unsigned int rest_val;
-  unsigned int rest_test;
+  void setStateAndNotify(const DataStatus status);
 
-  /**
-   * @brief     batch size
-   */
-  unsigned int batch_size;
+  DatasetType data_buffer_type; /**< data buffer type */
 
-  /**
-   * @brief     flags to check status
-   */
-  bool train_running;
-  bool val_running;
-  bool test_running;
+  /** data queue and producer/consumer status related variables */
+  std::thread
+    batch_producer;        /** thread generates a single batches to the queue */
+  std::mutex data_lock;    /**< data queue mutex */
+  std::mutex status_mutex; /**< data status mutex */
+  std::condition_variable cv_status; /**< condition variable for status */
+  std::vector<std::vector<float>> data_q;
+  std::vector<std::vector<float>> label_q;
+  DataStatus queue_status;  /**< status of the queue */
+  unsigned int buf_size;    /**< queue size */
+  unsigned int cur_bufsize; /**<  number of data in the data queue */
 
-  /**
-   * @brief     ids to check duplication
-   */
-  std::vector<unsigned int> train_mark;
-  std::vector<unsigned int> val_mark;
-  std::vector<unsigned int> test_mark;
+  TensorDim input_dim;     /**< feature size */
+  unsigned int class_num;  /**< number of class */
+  unsigned int batch_size; /**< batch size */
 
-  /**
-   * @brief     threads to generate data for queue
-   */
-  std::thread train_thread;
-  std::thread val_thread;
-  std::thread test_thread;
+  /**< @todo below variable should be owned by databuffer that has fixed size */
+  unsigned int samples_per_epoch; /**< size of samples in the dataset */
+  unsigned int remaining_samples_per_epoch; /**< size of samples remaining in
+                                               current epoch */
 
-  std::vector<std::string> labels;
-  bool validation[NBUFTYPE];
+  bool is_running; /**< flag to check if running */
+  bool is_valid;   /**< check if current buffer is in valid state */
```

Comment:
how about is_valid -> initialized?

Reply:
done. fyi) in the upcoming design this will not be used :)

Reply:
yes, that will be even better.